Fix Auto Functionalize to handle specified default values

flaviotruzzi opened the pull request:
Summary: When there were optionals with specified default values the code was improperly handling the number of parameters causing `IndexError: tuple index out of range`

Test Plan: new tests

Differential Revision: D52977644




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

flaviotruzzi commented:
@pytorchbot merge

yanboliang commented:
@pytorchbot merge

malfet commented on test/dynamo/test_misc.py:
```diff
@@ -427,6 +427,33 @@ def h(x):
+
+        @torch.library.impl("mylib::foo", "cpu", lib=lib)
+        def foo_impl(a, b, c=None, d=None, e=-1):
+            a + mode
```
Is this supposed to fail to compile?

flaviotruzzi replied:
Rebasing mistake, waiting your change to force run the tests before fixing it.

malfet commented:
https://github.com/pytorch/pytorch/pull/118035/commits/a245abc34f3feeb51b4f3757e6cdcbb7621bc47b should fails with `NameError`, but CI is green, slapped `ciflow/inductor` to see if that would make any difference

flaviotruzzi committed:
[Pytorch] - Fix Auto Functionalize to handle specified default values (#118035)

Summary:

When there were optionals with specified default values the code was improperly handling the number of parameters causing `IndexError: tuple index out of range`

Test Plan: new tests

Reviewed By: williamwen42

Differential Revision: D52977644

zou3519 commented:
@pytorchbot revert -m "needs internal changes, reverting so we can land via co-dev" -c weird

flaviotruzzi commented:
This landed on https://github.com/pytorch/pytorch/pull/118331
